bigquery-connector-common/src/main/java/com/google/cloud/bigquery/connector/common/BigQueryDirectDataWriterHelper.java

```diff
@@ -184,19 +194,81 @@ public void addRow(ByteString message) throws IOException {
     appendRequestRowCount++;
   }
 
+  /**
+   * Thorws an exception if there is an error in any of the responses received thus far. Since
+   * responses arrive in order, we proceed to check the next response only after the previous
+   * response has arrived.
+   *
+   * @throws BigQueryConnectorException If there was an invalid response
+   */
+  private void checkForFailedResponse(boolean waitForResponse) {
+    while (nextIndexCheckResponse < appendRowsFutures.size()) {
+      ApiFuture<AppendRowsResponse> validatedAppendRowsFuture =
+          appendRowsFutures.get(nextIndexCheckResponse);
```

Comment:
Can use Queue::peek here, and Queue::remove when we do nextIndexCheckResponse++

Reply:
I use peek() to retrieve and check whether head element is ready. Then I use poll() to remove the head.